feat(git_status): add variables to track worktree and index changes

mcwarman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->

Add additional variables to `github_status` to allow for a more granular view in the prompt.

Takes slight inspiration from #3276 but tries to do it in away that preserves backwards compatibility, by only making options available to those who chose to configure it. Also making sure none of the existing functionality is changed.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #3048

#### Screenshots (if appropriate):

![image](https://github.com/starship/starship/assets/7236500/66056794-3725-465c-a35d-a9f8859ad6b4)

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

<details><summary>Config</summary>
<p>

```toml
[git_status]
style = "yellow"
format = '([\[$ahead_behind( ($renamed )($index_added )($index_modified )($index_deleted )($index_typechanged )($worktree_added )($worktree_modified )($worktree_deleted )($worktree_typechanged ))\]( $stashed)]($style) )'
up_to_date = "[✓](blue)"
ahead = "⇡${count}"
diverged = "⇡${ahead_count}⇣${behind_count}"
behind = "⇣${count}"
staged = ""
typechanged = ""
modified = ""
deleted = ""
untracked = ""
renamed = "[»$count](green)"
index_added = "[+$count](green)"
index_modified = "[~$count](green)"
index_deleted = "[-$count](green)"
index_typechanged = "[⇢$count](red)"
worktree_added = "[+$count](red)"
worktree_modified = "[~$count](red)"
worktree_deleted = "[-$count](red)"
worktree_typechanged = "[⇢$count](red)"
untracked_as_worktree_added = true
stashed = "≡"
``` 

</p>
</details>

- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

mcwarman commented:
Hello @matchai

I'd be interested on your opinion on this as you previously commented on #3276, I think I've addressed the concerns raised.

mcwarman commented:
Hello @davidkna (sorry to `@`, but you seem most active in the review space)

I'd be interested in getting some feedback on this PR, what are your thoughts?

taketwo commented:
@matchai @davidkna This PR addresses a long-standing bug in one of the essential modules in a backward-compatible way. It comes with comprehensive test coverage and documentation updates--a solid contribution. Could you suggest the steps we need to take to get this merged?

brian6932 commented:
`git status` also has a `--show-stash` param, would be nice to get the stash count from there too (would make starship have git feature parity with oh-my-posh).

davidkna commented on src/modules/git_status.rs:
```diff
@@ -290,30 +399,71 @@ struct RepoStatus {
     fn is_deleted(short_status: &str) -> bool {
         // is_wt_deleted || is_index_deleted
-        short_status.contains('D')
+        Self::is_worktree_deleted(short_status) || Self::is_index_deleted(short_status)
```
Please just add the counts as needed for the regular variables.

mcwarman replied:
Thanks for the review.

Can you clarify what you're asking for here, the existing counts where left un-touched:

https://github.com/starship/starship/blob/c8d74b101f81bb0c50615e00106ebabb4a0552ad/src/modules/git_status.rs#L468-L482

Are you asking for me to leave the logic as it was? Therefore reverting changes made to `is_deleted`, `is_modified`, `is_staged` and `is_typchanged`?

Or are you asking for something else? 

The changes where just an attempt to refactor, but will function the same as far as I understand.

davidkna replied:
As far as I can tell, deleted and other variables are equal to e.g. `is_wt_deleted + is_index_deleted`, in that case I fell like it makes sense to keep track of these for each case and add these for the main `deleted` variable.

mcwarman replied:
Fixed in: https://github.com/starship/starship/pull/5655/commits/44b9dd8e75f85707e13860db376f3d098d7d85b1

mcwarman replied:
and then made clearer again here: https://github.com/starship/starship/pull/5655/commits/75b1e1b57ff02e5366b4175b002571f46dea2572

davidkna commented:
> `git status` also has a `--show-stash` param, would be nice to get the entries count from there too (would make starship have git feature parity with oh-my-posh).

Stash count is already supported.

brian6932 commented:
With what variable? `stashed` only shows if a stash is there or not, doesn't show how many stashes I have.

mcwarman commented on docs/config/README.md:
```diff
@@ -1967,43 +1967,60 @@ You can disable the module or use the `windows_starship` option to use a Windows
+| `index_deleted`               | `""`                                          | The format of `index_deleted`                                                                               |
+| `index_modified`              | `""`                                          | The format of `index_modified`                                                                              |
+| `index_typechanged`           | `""`                                          | The format of `index_typechanged`                                                                           |
+| `untracked_as_worktree_added` | `false`                                       | Add `untracked` files to `worktree_added` count                                                             |
```
I can remove it as a feature.

It was a attempt to create feature parity with how other options worked, by giving the user the option to control it.

davidkna replied:
I stand by my original comment.

mcwarman replied:
Fixed in https://github.com/starship/starship/pull/5655/commits/393410691ee67c9e57aac6b2c06939b23cca0f0d

mcwarman replied:
and then made clearer again here: https://github.com/starship/starship/pull/5655/commits/75b1e1b57ff02e5366b4175b002571f46dea2572

mcwarman commented:
@davidkna I believe I've addressed the review comments.

There was also a schema issue on master which was causing the CI to fail, I added the update in separate commit (https://github.com/starship/starship/pull/5655/commits/6c9dd8c4d63dccc2ee10ff44ae3e4bcf3566a3a4) , but it can be dropped if another merge resolves it.

mcwarman commented:
It looks like i've broken it in latest update please hold off reviewing sorry.

mcwarman commented on src/modules/git_status.rs:
```diff
@@ -290,46 +381,86 @@ struct RepoStatus {
+            self.index_typechanged += 1;
+        }
+
+        self.deleted = self.worktree_deleted + self.index_deleted;
+        self.modified = self.worktree_modified + self.worktree_added;
+        self.staged = self.index_modified + self.index_added + self.index_typechanged;
+        self.typechanged = self.worktree_typechanged;
```
Let me know if this should be moved to the end of the `add()` method instead.

mcwarman commented:
@davidkna Any other review comments?

mcwarman commented:
@davidkna 

Hello, hope you're well, before I go through and resolve these conflicts is there an appetite to merge this or should I just close it?

brian6932 commented:
> is there an appetite

I've been running a local rebased branch from this PR for a while now, so take that as you will.
